### PR TITLE
Fallback apply on and skip on 9671

### PR DESCRIPTION
**Pull request by Joseph-Cass**

### CODE REVIEW

**Comment by Joseph-Cass** [comment]
#build

File: dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/FallbackConfig21Impl.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.eclipse.microprofile.faulttolerance.Fallback;
+
+import com.ibm.ws.microprofile.faulttolerance.cdi.config.impl.FallbackConfigImpl;
+import com.ibm.ws.microprofile.faulttolerance.spi.FallbackPolicy;
+
+/**
+ *
+ */
+public class FallbackConfig21Impl extends FallbackConfigImpl {
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.1.cdi/src/com/ibm/ws/microprofile/faulttolerance21/cdi/config/impl/FallbackConfig21Impl.java]: In general, either write something in the javadoc comment or remove it completely

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FallbackConfig.java [hunk 2]
```diff
@@ -10,197 +10,21 @@
...
 
-    private static FallbackPolicy newFallbackPolicy(Class<? extends FallbackHandler<?>> fallbackHandlerClass, FallbackHandlerFactory factory) {
-        FallbackPolicy fallbackPolicy = FaultToleranceProvider.newFallbackPolicy();
-        fallbackPolicy.setFallbackHandler(fallbackHandlerClass, factory);
-        return fallbackPolicy;
-    }
+    /**
+     * Validate the CircuitBreaker policy
+     */
+    void validate();
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FallbackConfig.java]: Correct this comment

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FallbackConfig.java [hunk 3]
```diff
@@ -10,197 +10,21 @@
...
-            }
-        };
-
-        FallbackPolicy fallbackPolicy = newFallbackPolicy(fallbackFunction);
-        return fallbackPolicy;
-    }
+/**
+ *
+ */
+public interface FallbackConfig {
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/FallbackConfig.java]: Either write a comment here or remove the empty javadoc

File: dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/FallbackConfigImpl.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.cdi/src/com/ibm/ws/microprofile/faulttolerance/cdi/config/impl/FallbackConfigImpl.java]: I think this file was just moved? If so it should still have the original copyright date. Should be `2017, 2019`.

File: dev/com.ibm.ws.microprofile.faulttolerance.spi/src/com/ibm/ws/microprofile/faulttolerance/spi/FallbackPolicy.java [hunk 5]
```diff
@@ -24,4 +24,25 @@
...
+
+    @SuppressWarnings("unchecked")
+    public void setApplyOn(Class<? extends Throwable>... failOn);
+
+    /**
+     * Define the skip criteria
+     *
+     * @return the skip exception
+     */
+    Class<? extends Throwable>[] getSkipOn();
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.spi/src/com/ibm/ws/microprofile/faulttolerance/spi/FallbackPolicy.java]: Add `public` here for consistency with the other methods in the class.

File: dev/com.ibm.ws.microprofile.faulttolerance.2.0/src/com/ibm/ws/microprofile/faulttolerance20/state/FallbackState.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public boolean shouldApplyFallback(MethodResult<?> result);
+
+    /**
+     * Executes the fallback function based on the executionContext
+     *
+     * @param <R>              the return type of the method
+     * @param result           the result of the execution of the method Fallback was applied to
+     * @param executionContext context for the execution of the method Fallback was applied to
+     * @return an updated MethodResult with details on whether Fallback was applied or not
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.0/src/com/ibm/ws/microprofile/faulttolerance20/state/FallbackState.java]: The returned `MethodResult` doesn't provide details on whether or not Fallback was applied, it just returns the new result obtained by running fallback.

File: dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+public class FallbackStateImplTest {
+
+    private static MetricRecorder dummyMetrics = DummyMetricRecorder.get();
+
+    @Test
+    public <R> void testFallback() throws InterruptedException, InstantiationException, IllegalAccessException {
+        FallbackPolicyImpl policy = new FallbackPolicyImpl();
+        FallbackStateImpl state = new FallbackStateImpl(policy, dummyMetrics);
+
+        MethodResult<R> resultWithException = MethodResult.failure(RuntimeException.class.newInstance());
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java]: Replace `RuntimeException.class.newInstance()` with `new RuntimeException()`

File: dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        assertTrue("Fallback should be applied if an exception is thrown, according to Fallback's default applyOn() and skipOn() values",
+                   state.shouldApplyFallback(resultWithException));
+
+        MethodResult<R> resultNoException = MethodResult.success(null);
+        assertFalse("Fallback should NOT be applied if an exception is NOT thrown, according to Fallback's default applyOn() and skipOn() values",
+                    state.shouldApplyFallback(resultNoException));
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java]: Could make these assertion messages shorter. Just "Fallback should (NOT) be applied if an exception is (NOT) thrown" would cover what the test is checking.

File: dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public <R> void testFallbackApplyOn() throws InterruptedException, InstantiationException, IllegalAccessException {
+        FallbackPolicyImpl policy = new FallbackPolicyImpl();
+        policy.setApplyOn(TestExceptionA.class);
+        FallbackStateImpl state = new FallbackStateImpl(policy, dummyMetrics);
+
+        MethodResult<R> resultExceptionA = MethodResult.failure(TestExceptionA.class.newInstance());
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java]: Change all these occurences of `newInstance()` to `new TestExceptionBlah()`

File: dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public <R> void testFallbackSkipOn() throws InterruptedException, InstantiationException, IllegalAccessException {
+        FallbackPolicyImpl policy = new FallbackPolicyImpl();
+        policy.setSkipOn(TestExceptionA.class);
+        FallbackStateImpl state = new FallbackStateImpl(policy, dummyMetrics);
+
+        //
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.faulttolerance.2.0/test/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/FallbackStateImplTest.java]: Take out the empty comment lines on this method

**Comment by Joseph-Cass** [comment]
#build

**Comment by Azquelt** [comment]
#run-libby-bot

**Pull request merged**